core/embed/rust/src/ui/shape/qrcode.rs

```diff
@@ -0,0 +1,169 @@
+}
+
+impl QrImage {
+    pub fn new(area: Rect, qrcode: &QrCode) -> Self {
+        if area.width() < qrcode.size() as i16 || area.height() < qrcode.size() as i16 {
+            panic!("Too small area");
+        }
+
+        let mut result = QrImage {
+            area,
+            qr_size: qrcode.size() as i16,
+            qr_modules: [0u8; MAX_QRCODE_BYTES],
+            fg_color: Color::white(),
+            bg_color: None,
+        };
+
+        // Copy content of QR code to the qrmodules buffer
+        for y in 0..result.qr_size {
+            for x in 0..result.qr_size {
+                result.set_module(x, y, qrcode.get_module(x as i32, y as i32));
+            }
+        }
+
+        result
+    }
+
+    fn set_module(&mut self, x: i16, y: i16, value: bool) {
+        // Every row starts at byte aligned address
+        let row_offset = (y * (self.qr_size + 7) / 8) as usize;
+        let row = &mut self.qr_modules[row_offset..];
+        let col_offset = (x / 8) as usize;
+        let col_bit = 1 << (x & 0x7);
+        if value {
+            row[col_offset] |= col_bit;
+        } else {
+            row[col_offset] &= col_bit ^ 0xFF;
+        }
+    }
+
+    fn get_module(&self, x: i16, y: i16) -> bool {
+        // Every row starts at byte aligned address
+        let row_offset = (y * (self.qr_size + 7) / 8) as usize;
+        let row = &self.qr_modules[row_offset..];
+        let col_offset = (x / 8) as usize;
+        let col_bit = 1 << (x & 0x7);
+        (row[col_offset] & col_bit) != 0
+    }
+
+    pub fn with_fg(self, fg_color: Color) -> Self {
+        Self { fg_color, ..self }
+    }
+
+    pub fn with_bg(self, bg_color: Color) -> Self {
+        Self {
+            bg_color: Some(bg_color),
+            ..self
+        }
+    }
+
+    pub fn render<'s>(self, renderer: &mut impl Renderer<'s>) {
+        renderer.render_shape(self);
+    }
+
+    fn draw_row(&self, slice_row: &mut [u8], qr_y: i16) {
+        slice_row.iter_mut().for_each(|b| *b = 0);
+
+        let mut qr_module = false;
+
+        for p in line_points(self.area.x1 - self.area.x0, self.qr_size, 0) {
+            if p.first {
+                qr_module = self.get_module(p.v, qr_y);
+            }
+            if !qr_module {
+                if p.u & 0x01 == 0 {
+                    slice_row[(p.u / 2) as usize] |= 0x0F;
+                } else {
+                    slice_row[(p.u / 2) as usize] |= 0xF0;
+                }
+            }
+        }
+    }
+}
+
+impl Shape<'_> for QrImage {
+    fn bounds(&self) -> Rect {
+        self.area
+    }
+
+    fn cleanup(&mut self, _cache: &DrawingCache) {}
+
+    fn draw(&mut self, canvas: &mut dyn Canvas, cache: &DrawingCache) {
+        let buff = &mut unwrap!(cache.image_buff(), "No TOIF buffer");
+
+        let mut slice = unwrap!(
+            Bitmap::new_mut(
+                BitmapFormat::MONO4,
+                None,
+                Offset::new(self.area.width(), 1),
+                Some(1),
```

Comment:
are we explicitly requesting height 1 here? then we should not be setting min_height right?

Reply:
Yes, your are right. This is unnecessary. See fix a1dece7a009ec1619fd9dacf3efe57435c51740c